content/en/_index.md

```diff
@@ -478,6 +583,14 @@ try {
   // IllegalArgumentException's are thrown when an invalid argument has been passed to a function. Ok to ignore.
 }
 {{< /tab >}}
+{{< tab header="Go" >}}
+// If the sample can run into errors, write the error message to the provided
+// io.Writer. Don't call log.Fatal or friends.
+if err != nil {
+    fmt.Fprintf(w, "foo.NewClient: %v", err)
```

Comment:
If we follow the text above, this is what the pattern becomes in Go, which is not idiomatic Go, unlike throwing exceptions in Java, Python, Ruby, and NodeJS.

Reply:
If the idea is that we alert users to doing something with the error, maybe this is OK, but typically, Go developers would return an error.

Reply:
My 2cents is that logging the error to console is what we want to do here. Logging the error as opposed to just returning it makes it easier to run on the user.

Reply:
I've thought about this as well and come to the same conclusion. Go developers know how to return errors. I think the idea of the samples here is more about demonstrating usage. I've added a `TODO(developer): handle error` just below to communicate that. Otherwise, writing to the error message to the `io.Writer` will communicate what went wrong, so that when people copy-paste and run the sample, they get clear feedback on what went wrong.

@codyoss You're one of the Go stakeholders here. How does this sound?

Reply:
This looks fine to me - my two nits are:

1. I think we should lead with the comment to clarify the user should update this section as needed. 
2. I think we should push for more clear (even if a bit more verbose) messages  since these are samples. 

e.g:

```
if err != nil {
    // TODO(developer): return or handle error as necessary  
    fmt.Fprintf(w, "failed to initialize foo.NewClient: %v", err)
   return
}
```

Reply:
If we want to log the errors I guess that is fine, but for Go I think we should definitely still return the error as well as a part of the signature. The rest of the programing languages we support operate on an exception paradigm(I believe), and Go does not. Errors are just values and Go and should be handled. I think if we had samples that did not return errors that sets a bad precedence for our users. If this controversial take I am happy to take it offline and we can chat in a better forum. But I have a strong option here that we should return errors.

Reply:
Here's my elevator pitch of trying to convince you why this rule was made in the OG 2019 samples working group. Happy to set up a time to chat as well if a higher bandwidth forum is needed. 

I do not think this is radically different than what is happening in other languages. Java specifically has 2 categories of exceptions: checked and unchecked exceptions. There are equivalent to a return value (it has to be added to the function signature) or panic (not in the function signature). 

Based on UX testing, one of the biggest goals was to create samples that are "copy-paste-runnable". The reason the rule came about around forbidding return values is because there was a pattern of snippets that create some object and then return it to be inspected for tests. These were less useful to users because it often showed the initialization of the object but not actually how to use the object in any meaningful way. It also made it harder to tell if it ran successfully, because now the user has to do something with those values to actually tell if the sample was successful or not. So the rule was created to "don't return, output to console" to force the result of the function (successful or otherwise).

I agree with you printing errors to console is not idiomatic to Go (it's not idiomatic to any of the languages). I also agree that we should be try to be handling errors in the samples (when it makes sense). However, returning an error is not handling. It's just moving the handling of the error to a different part of the code, likely not in the snippet (outside of what the user sees). But sometimes there isn't anything in the program to handle for errors - the application needs to crash and the user needs to take some non-programmatic action outside of this (enable API, change file permissions, etc). By putting a comment and writing the error to console, we are doing the following: 
1. The TODO makes it clear this isn't an idiomatic practice, and the developer should do what makes sense for their application
2. It helps highlight the cases where the error _should_ be handled in the sample 
3. It streamlines running the sample by forcing all results to the same place - the user can expect to see _something_ in the output from the function and can't just ignore the return value (accidently or otherwise) 
4. It also forces the test to actually check the output of the sample, which is the same way users are determining if the function is successful or not. 

So in conclusion, I don't think this pattern is perfect, but I think it's better than just returning values to somewhere else because it encourages better samples overall.

Reply:
I see in one of the Java examples in this repo that is has a throws clause. To me this is similarish to returning an error in Go as we are asking the caller to due something about it. But returning an error is much lighter weight than declaring a function can throw an error. The code is just as copy-pastable if we return an error -- the caller does not even have to create a variable to capture the error or do anything with it in Go. I understand this stance of printing to show how to use for API responses, but to me errors are different, and if we are not doing any introspection with them we might as well return them as well. It makes the code more "production ready" from the get-go.

Reply:
> I see in one of the Java examples in this repo that is has a throws clause.

[Here is the specific rule for Java](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md#exception-handling), which here is an exception that we only let low level environment/hardware errors in the signature (again, largely because historically Java has done a poor job with Runtime Exceptions vs checked Exceptions). It specifically says all other errors should be handled or logged in the sample. 

> The code is just as copy-pastable if we return an error -- the caller does not even have to create a variable to capture the error or do anything with it in Go

A key part is samples should be copy-paste-_runnable_. Ideally samples should be easy to run as well, and leaving a return type up to the user means that could just ignore/miss it and not have feedback on why the sample failed. Returning the error means they just have to write the `if err != nil fmt.Println(err)` themselves. (Tangentially, this is why I think it's a bad idea to use a writer over fmt.Print - it's just adding extra layers to actually running the sample - but that's another argument)

> It makes the code more "production ready" from the get-go.

This isn't production code. It's meant to be instructional. We already follow a bunch of practices that don't happen in production code -  we don't typically create a new client for each request or use a writer to output content either. IMO, the comment disclaimer here clarifies that this isn't a production practice. This is why we put a comment on client initialization as well.

Reply:
For what it's worth, my own view is that returning errors is a well-established pattern that any reasonably competent Go developer knows well. So I assume anyone seeing an error printed to stdout with a note above it that says "handle or return error," will understand that the sample is about demonstrating usage and will adjust the error handling to their own needs.

cc @tritone

Reply:
I'll +1 @codyoss 's response here. Not having an error returned in Go is, well, strange. It's easier to start adapting a function from a sample code to a production use if the signature is closer to what will be correct, and it also makes the sample code more testable. Also, languages other than Java (e.g. python) will always bubble up errors as exceptions unless there is a try/except block in every sample, which I sincerely hope we're not planning on.

I also question whether the style guide needs to take a firm stance on this one way or the other. What is the cost of having this be up to whoever is implementing the samples, especially given that existing samples generally return an error and we don't plan on migrating them to follow the guide?

Reply:
Removing established rules is not in scope right now, but it's clear this is one of the rules that hits one of Go's strongest language opinions. I think we should go to the goals of the rule, figure out what's needed, then consider if the rule's language might need revisions.

The main goal as I understand is to make sure that **developers handle errors and avoid failures that a learner can't distinguish from a product error**

With that goal in mind, and in the background other concerns about copy-paste-runnable and cross-language consistency, I think it's fair to reframe the strict expectations we have in other languages.

Looking at this again, my thoughts are:
* Returning an error in go is idiomatic, not returning an error is almost never ok
* For Go developers, returning the error does not seem like an excessive distraction
* Not sure why I'd log an error to an io.Writer instead of, well, logging the error. This makes sense to me for catching function output, not errors
* We shouldn't assume sample readers or authors are experts in Go, programming, or Cloud.

Could we remind folks to log, have the copied sample log by default, and idiomatically return the error with something like this?

```
// If the sample can run into errors, write the error message to the provided
if err != nil {
    // TODO(developer): log this error from the calling function.
    log.Println("failed to initialize foo.Client: %v", err)
    return err
}
```

If this proposal isn't getting us closer to a resolution, let's plan to talk this through.

Reply:
To summarize an offline conversation with @tritone, @codyoss, @kurtisvg, we're going to keep the error return for now. At some future date, we will remove the io.Writer, work that depends on a clear testing strategy.